plugins/redis/redis.cpp

```diff
@@ -0,0 +1,904 @@
+{
+    StringArray optionStrings;
+    optionStrings.appendList(_options, " ");
+    ForEachItemIn(idx, optionStrings)
+    {
+        const char *opt = optionStrings.item(idx);
+        if (strncmp(opt, "--SERVER=", 9) == 0)
+        {
+            opt += 9;
+            StringArray splitPort;
+            splitPort.appendList(opt, ":");
+            if (splitPort.ordinality()==2)
+            {
+                ip.set(splitPort.item(0));
+                port = atoi(splitPort.item(1));
+            }
+        }
+        else
+        {
+            VStringBuffer err("RedisPlugin: unsupported option string %s", opt);
+            rtlFail(0, err.str());
+        }
+    }
+    if (ip.isEmpty())
+    {
+        ip.set("localhost");
+        port = 6379;
+        if (ctx)
+        {
+            VStringBuffer msg("Redis Plugin: WARNING - using default server (%s:%d)", ip.str(), port);
+            ctx->logString(msg.str());
+        }
+    }
+    return;
+}
+void Connection::authenticate(ICodeContext * ctx, const char * password)
+{
+    if (password && *password)
+    {
+        OwnedReply reply = Reply::createReply(redisCommand(context, "AUTH %b", password, strlen(password)));
+        assertOnError(reply->query(), "server authentication failed");
+    }
+}
+void Connection::resetContextErr()
+{
+    if (context)
+        context->err = REDIS_OK;
+}
+void Connection::readReply(Reply * reply)
+{
+    redisReply * nakedReply = NULL;
+    redisGetReply(context, (void**)&nakedReply);
+    assertex(reply);
+    reply->setClear(nakedReply);
+}
+void Connection::readReplyAndAssert(Reply * reply, const char * msg)
+{
+    readReply(reply);
+    assertex(reply);
+    assertOnError(reply->query(), msg);
+}
+void Connection::readReplyAndAssertWithKey(Reply * reply, const char * msg, const char * key)
+{
+    readReply(reply);
+    assertex(reply);
+    assertOnCommandErrorWithKey(reply->query(), msg, key);
+}
+Connection * Connection::createConnection(ICodeContext * ctx, const char * options, unsigned __int64 _database, const char * password, unsigned __int64 _timeout)
+{
+    if (!cachedConnection)
+    {
+        cachedConnection = new Connection(ctx, options, _database, password, _timeout);
+        threadHookChain = addThreadTermFunc(releaseContext);
+        return LINK(cachedConnection);
+    }
+
+    if (cachedConnection->isSameConnection(ctx, password))
+    {
+        //MORE: should perhaps check that the connection has not expired (think hiredis REDIS_KEEPALIVE_INTERVAL is defaulted to 15s).
+        //At present updateTimeout calls assertConnection.
+        cachedConnection->resetContextErr();//reset the context err to allow reuse when an error previously occurred.
+        cachedConnection->updateTimeout(_timeout);
+        cachedConnection->selectDB(ctx, _database);
+        return LINK(cachedConnection);
+    }
+
+    cachedConnection->Release();
+    cachedConnection = new Connection(ctx, options, _database, password, _timeout);
```

Comment:
Should set cachedConnection to NULL here.  Otherwise if new Connection fails it will not update cachedConnection, and subsequent calls will access invalid memory.